File: java/src/main/java/TennisGame1.java

public class TennisGame1 implements TennisGame {

    private int player1Score = 0;
    private int player2Score = 0;
    private String player1;
    private String player2;
    private static final String[] POINTS = {"Love", "Fifteen", "Thirty", "Forty"};
    private static final int WIN_THRESHOLD = 4;

    public TennisGame1(String player1, String player2) {
        this.player1 = player1;
        this.player2 = player2;
    }

    public void wonPoint(String player) {
        if (player.equals(player1))
            player1Score += 1;
        else
            player2Score += 1;
    }

    public String getScore() {
        if (isDeuce()) {
            return "Deuce";
        } else if (isAdvantageOrWin()) {
            return getAdvantageOrWinScore();
        } else if (player1Score == player2Score) {
            return POINTS[player1Score] + "-All";
        } else {
            return POINTS[player1Score] + "-" + POINTS[player2Score];
        }
    }

    private boolean isDeuce() {
        return player1Score >= 3 && player1Score == player2Score;
    }

    private boolean isAdvantageOrWin() {
        return player1Score >= WIN_THRESHOLD || player2Score >= WIN_THRESHOLD;
    }

    private String getAdvantageOrWinScore() {
        int scoreDifference = player1Score - player2Score;
        if (scoreDifference == 1) {
            return "Advantage player1";
        } else if (scoreDifference == -1) {
            return "Advantage player2";
        } else if (scoreDifference >= 2) {
            return "Win for player1";
        } else {
            return "Win for player2";
        }
    }
}

File: typescript/src/TennisGame1.ts
import { TennisGame } from './TennisGame'

class Player {
  constructor(public name: string, public score: number = 0) {}
}

export class TennisGame1 implements TennisGame {
  private player1: Player
  private player2: Player

  constructor(player1Name: string, player2Name: string) {
    this.player1 = new Player(player1Name)
    this.player2 = new Player(player2Name)
  }

  wonPoint(playerName: string): void {
    if (playerName === 'player1') this.player1.score += 1
    else this.player2.score += 1
  }

  getScore(): string {
    if (this.isDraw()) {
      return this.scoreWhenDraw()
    }
    if (this.oneOfTheTwoScoredForty()) {
      return this.calculateWinnerOrAdvantage()
    }

    return this.nonWinningScore()
  }

  private scoreWhenDraw() {
    if (this.player1.score < 3) {
      return this.getScoreName(this.player1.score) + '-All'
    } else {
      return 'Deuce'
    }
  }

  private oneOfTheTwoScoredForty() {
    return this.player1.score >= 4 || this.player2.score >= 4
  }

  private isDraw() {
    return this.player1.score === this.player2.score
  }

  private nonWinningScore() {
    return (
      this.getScoreName(this.player1.score) +
      '-' +
      this.getScoreName(this.player2.score)
    )
  }

  private getScoreName(score: number) {
    switch (score) {
      case 0:
        return 'Love'
      case 1:
        return 'Fifteen'
      case 2:
        return 'Thirty'
      case 3:
        return 'Forty'
      default:
        return ''
    }
  }

  private calculateWinnerOrAdvantage(): string {
    const scoreDifference = Math.abs(this.player1.score - this.player2.score)
    const winningPlayer =
      this.player1.score > this.player2.score ? this.player1 : this.player2

    if (scoreDifference === 1) return 'Advantage ' + winningPlayer.name
    if (scoreDifference >= 2) return 'Win for ' + winningPlayer.name
  }
}
